sw/device/lib/crypto/aes_gcm/aes_gcm.c

```diff
@@ -0,0 +1,351 @@
+    0x54a4, 0x96a5, 0xd8a0, 0x1aa1, 0x5ca3, 0x9ea2, 0xe0b5, 0x22b4, 0x64b6,
+    0xa6b7, 0xe8b2, 0x2ab3, 0x6cb1, 0xaeb0, 0xf0bb, 0x32ba, 0x74b8, 0xb6b9,
+    0xf8bc, 0x3abd, 0x7cbf, 0xbebe};
+
+/**
+ * Performs a bitwise XOR of two blocks.
+ *
+ * This operation corresponds to addition in the Galois field.
+ *
+ * @param condition Either 0 or 1
+ * @param x First operand block
+ * @param y Second operand block
+ * @param out Buffer in which to store output; can be the same as one or both
+ * operands.
+ */
+static inline void block_xor(const aes_block_t *x, const aes_block_t *y,
+                             aes_block_t *out) {
+  for (size_t i = 0; i < kAesBlockNumWords; i++) {
+    out->data[i] = x->data[i] ^ y->data[i];
+  }
+}
+
+/**
+ * Reverse the bits of a byte.
+ *
+ * @param word Input word.
+ * @return Word with bytes reversed.
+ */
+static uint8_t reverse_bits(uint8_t byte) {
+  uint8_t out = 0;
+  for (size_t i = 0; i < 8; i++) {
+    out <<= 1;
+    out |= (byte >> i) & 1;
+  }
+  return out;
+}
+
+/**
+ * Reverse the bytes of a 32-bit word.
+ *
+ * This function is convenient for converting between the processor's native
+ * integers, which are little-endian, and NIST's integer representation, which
+ * is big-endian.
+ *
+ * @param word Input word.
+ * @return Word with bytes reversed.
+ */
+static uint32_t reverse_bytes(uint32_t word) {
+  uint8_t *word_bytes = (uint8_t *)&word;
+  uint32_t out;
+  for (size_t i = 0; i < sizeof(uint32_t); i++) {
+    out <<= 8;
+    out |= word_bytes[i];
+  }
+  return out;
+}
+
+/**
+ * Logical right shift of an AES block.
+ *
+ * NIST represents AES blocks as big-endian, and a "right shift" in that
+ * representation means shifting out the LSB. However, the processor is
+ * little-endian, so we need to reverse the bytes before shifting.
+ *
+ * @param block Input AES block, modified in-place
+ * @param nbits Number of bits to shift
+ */
+static inline void block_shiftr(aes_block_t *block, size_t nbits) {
+  // To perform a "right shift" with respect to the big-endian block
+  // representation, we traverse the words of the block and, for each word,
+  // reverse the bytes, save the new LSB as `carry`, shift right, set the MSB
+  // to the last block's carry, and then reverse the bytes back.
+  uint32_t carry = 0;
+  for (size_t i = 0; i < kAesBlockNumWords; i++) {
+    uint32_t rev = reverse_bytes(block->data[i]);
+    uint32_t next_carry = rev & ((1 << nbits) - 1);
+    rev >>= nbits;
+    rev |= carry << (32 - nbits);
+    carry = next_carry;
+    block->data[i] = reverse_bytes(rev);
+  }
+}
+
+/**
+ * Retrieve the byte at a given index from the AES block.
+ *
+ * @param block Input AES block
+ * @param index Index of byte to retrieve (must be < `kAesBlockNumBytes`)
+ * @return Value of block[index]
+ */
+static inline uint8_t block_byte_get(const aes_block_t *block, size_t index) {
+  return ((uint8_t *)block->data)[index];
```

Comment:
To be very pedantic, isn't this a strict aliasing violation since `block->data` is an array of uint32_t and uint8_t is not guaranteed to be a char?

Reply:
I think so (and I think we might have other instances of this). How about:

```suggestion
  return (uint8_t)((char *)block->data)[index];
```

Reply:
I could also change all these bytes to `char` instead. I used `uint8_t` because I was modelling off of the pointer casts in `memory.c`: https://github.com/lowRISC/opentitan/blob/a96565ddbe684f14663c187db8a6c17908eebfbc/sw/device/lib/base/memory.c#L13-L20

I guess passing through `void *` avoids the strict aliasing issue?

Reply:
My understanding is that strict aliasing is the assumption that pointers of different types will never alias the same memory location. There are exceptions that allow for casting to character-type pointers, casting any `T *` to `const T *`, etc.

I don't believe passing through `void *` avoids the potential aliasing issue — the input pointers could have been something other than `uint8_t *`. If we're being very strict, we should use `char *` in the implementation of memcpy() as well.

However, if we're comfortable assuming that `uint8_t` is typedefed as `unsigned char` (and always will be), I guess my original point about strict aliasing is moot. In that case, it's more of a style issue.

Reply:
Okay, I've made the change that Alphan suggested here and also fixed similar places elsewhere in the code. Here's the diff (since I'm going to rebase it into a previous commit, so it won't be straightforward to view here); mind taking a quick look?
```diff
@@ -82,11 +82,10 @@ static uint8_t reverse_bits(uint8_t byte) {
  * @return Word with bytes reversed.
  */
 static uint32_t reverse_bytes(uint32_t word) {
-  uint8_t *word_bytes = (uint8_t *)&word;
   uint32_t out = 0;
   for (size_t i = 0; i < sizeof(uint32_t); i++) {
     out <<= 8;
-    out |= word_bytes[i];
+    out |= (word >> (8 * i)) & 255;
   }
   return out;
 }
@@ -125,7 +124,7 @@ static inline void block_shiftr(aes_block_t *block, size_t nbits) {
  * @return Value of block[index]
  */
 static inline uint8_t block_byte_get(const aes_block_t *block, size_t index) {
-  return ((uint8_t *)block->data)[index];
+  return (uint8_t)((char *)block->data)[index];
 }
 
 /**
@@ -311,9 +310,8 @@ static void aes_gcm_ghash_update(const aes_block_t H_table[16],
     if ((i == nblocks - 1) && (get_block_offset(input_len) != 0)) {
       // Last block is not full; pad with zeroes.
       size_t nbytes = get_block_offset(input_len);
-      uint8_t *input_data = (uint8_t *)input_block.data;
-      memcpy(input_data, input, nbytes);
-      memset(input_data + nbytes, 0, kAesBlockNumBytes - nbytes);
+      memset(input_block.data, 0, kAesBlockNumBytes);
+      memcpy(input_block.data, input, nbytes);
     } else {
       // Full block; copy data and advance input pointer.
       memcpy(input_block.data, input, kAesBlockNumBytes);
```
The last change is slightly less efficient than what I had before, but I don't think it's a big difference and this version is more readable anyway :slightly_smiling_face:

Reply:
> The last change is slightly less efficient than what I had before, but I don't think it's a big difference and this version is more readable anyway 🙂

Okay, I see it used to memcpy to `input_data[0..nbytes]` and then zero out the rest, but now it zeroes out the whole block before the memcpy. Was that motivated by the strict aliasing concern?

Other than that, the inline diff LGTM!

Reply:
> Okay, I see it used to memcpy to `input_data[0..nbytes]` and then zero out the rest, but now it zeroes out the whole block before the memcpy. Was that motivated by the strict aliasing concern?

The previous implementation casted the `uint32_t *` to a `uint8_t *` in order to make the pointer arithmetic work out for the second write (`input_data + nbytes`). But this was a little hacky and I'd been considering setting it all to zero before the copy anyway; the fact that the cast needed to make the pointer arithmetic nice violated strict aliasing just made the style/speed decision easier!

Reply:
Actually for the last chunk we can use:
```
memcpy(input_block.data, input, nbytes);
memset((char *)input_block.data + nbytes, 0, kAesBlockNumBytes - nbytes);
```